src/Lean/Elab/Tactic/Ext/Attr.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2021 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, Mario Carneiro
+-/
+import Lean.Elab.Tactic.RCases
+import Lean.Meta.DiscrTree
+import Lean.Elab.Command
+
+namespace Lean.Elab.Tactic.Ext
+open Lean Meta
+
+/-- `declare_ext_theorems_for A` declares the extensionality theorems for the structure `A`. -/
+syntax "declare_ext_theorems_for " ("(" &"flat" " := " term ") ")? ident (ppSpace prio)? : command
+
+/-- Information about an extensionality theorem, stored in the environment extension. -/
+structure ExtTheorem where
+  /-- Declaration name of the extensionality theorem. -/
+  declName : Name
+  /-- Priority of the extensionality theorem. -/
+  priority : Nat
+  /-- Key in the discrimination tree. -/
+  keys : Array DiscrTree.Key
+  deriving Inhabited, Repr, BEq, Hashable
+
+/-- The state of the `ext` extension environment -/
+structure ExtTheorems where
+  /-- The tree of `ext` extensions. -/
+  tree   : DiscrTree ExtTheorem := {}
+  /-- Erased `ext`s. -/
+  erased  : PHashSet Name := {}
+  deriving Inhabited
+
+/-- Discrimation tree settings for the `ext` extension. -/
+def extExt.config : WhnfCoreConfig := {}
+
+/-- The environment extension to track `@[ext]` lemmas. -/
+initialize extExtension :
+    SimpleScopedEnvExtension ExtTheorem ExtTheorems ←
+  registerSimpleScopedEnvExtension {
+    addEntry := fun { tree, erased } thm =>
+      { tree := tree.insertCore thm.keys thm, erased := erased.erase thm.declName }
+    initial := {}
+  }
+
+/-- Get the list of `@[ext]` lemmas corresponding to the key `ty`,
+ordered from high priority to low. -/
+@[inline] def getExtLemmas (ty : Expr) : MetaM (Array ExtTheorem) := do
+  let extTheorems := extExtension.getState (← getEnv)
+  let arr ← extTheorems.tree.getMatch ty extExt.config
+  let erasedArr := arr.filter fun thm => !extTheorems.erased.contains thm.declName
+  -- Using insertion sort because it is stable and the list of matches should be mostly sorted.
+  -- Most ext lemmas have default priority.
+  return erasedArr.insertionSort (·.priority < ·.priority) |>.reverse
+
+/-- Erases a name marked `ext` by adding it to the state's `erased` field and
+  removing it from the state's list of `Entry`s. -/
+def ExtTheorems.eraseCore (d : ExtTheorems) (declName : Name) : ExtTheorems :=
+ { d with erased := d.erased.insert declName }
+
+/--
+  Erase a name marked as a `ext` attribute.
```

Comment:
```suggestion
  Erases a name marked as a `ext` attribute.
```

It'd also be nice to include what erasure means in this context.

Reply:
Done.